/********************************************************************************
**
**	Copyright (C) 2007-2015 Witold Gantzke & Kirill Lepskiy
**
**	This file is part of the ACF-Solutions Toolkit.
**
**	This file may be used under the terms of the GNU Lesser
**	General Public License version 2.1 as published by the Free Software
**	Foundation and appearing in the file LicenseLGPL.txt included in the
**	packaging of this file.  Please review the following information to
**	ensure the GNU Lesser General Public License version 2.1 requirements
**	will be met: http://www.gnu.org/licenses/old-licenses/lgpl-2.1.html.
**
**	If you are unsure which license is appropriate for your use, please
**	contact us at info@imagingtools.de.
**
** 	See http://www.ilena.org or write info@imagingtools.de for further
** 	information about the ACF.
**
********************************************************************************/


#include <iblobgui/CBlobFilterGui.h>


// ACF includes
#include <iqt/CSignalBlocker.h>


namespace iblobgui
{


// public methods

CBlobFilterGui::CBlobFilterGui(QWidget* /*parentPtr*/)
{
	setupUi(this);

	ConditionCombo->setCurrentIndex(0);
	ParameterCombo->setCurrentIndex(0);

	connect(ParameterCombo, SIGNAL(currentIndexChanged(int)), this, SLOT(OnFilterParameterChanged()));
	connect(ConditionCombo, SIGNAL(currentIndexChanged(int)), this, SLOT(OnFilterParameterChanged()));
	connect(Value1SpinBox, SIGNAL(valueChanged(double)), this, SLOT(OnFilterParameterChanged()));
	connect(Value2SpinBox, SIGNAL(valueChanged(double)), this, SLOT(OnFilterParameterChanged()));

	UpdateComboBoxes();
}


void CBlobFilterGui::retranslateUi(QWidget *widgetPtr)
{
	int ConditionComboIndex = ConditionCombo->currentIndex();
	int ParameterComboIndex = ParameterCombo->currentIndex();

	Ui::CFilterWidget::retranslateUi(widgetPtr);

	UpdateComboBoxes();

	ConditionCombo->setCurrentIndex(ConditionComboIndex);
	ParameterCombo->setCurrentIndex(ParameterComboIndex);
}


void CBlobFilterGui::SetFilterInfo(const iblob::IBlobFilterParams::Filter& info)
{
	int featureIndex = ParameterCombo->findData(info.featureId);
	ParameterCombo->setCurrentIndex(featureIndex);

	int conditionIndex = ConditionCombo->findData(info.condition);
	ConditionCombo->setCurrentIndex(conditionIndex);

	Value1SpinBox->setValue(info.valueRange.GetMinValue());
	Value2SpinBox->setValue(info.valueRange.GetMaxValue());

	UpdateFilter();
}


iblob::IBlobFilterParams::Filter CBlobFilterGui::GetFilterInfo() const
{
	iblob::IBlobFilterParams::Filter info;

	info.featureId = ParameterCombo->itemData(ParameterCombo->currentIndex()).toByteArray();
	info.condition = iblob::IBlobFilterParams::FilterCondition(ConditionCombo->itemData(ConditionCombo->currentIndex()).toInt());
	info.operation = iblob::IBlobFilterParams::FO_INCLUDE;
	info.valueRange.SetMinValue(Value1SpinBox->value());
	info.valueRange.SetMaxValue(Value2SpinBox->value());

	return info;
}


void CBlobFilterGui::SetSupportedProperties(const iprm::IOptionsList& features)
{
	iqt::CSignalBlocker block(ParameterCombo);

	ParameterCombo->clear();

	int featuresCount = features.GetOptionsCount();
	for (int i = 0; i < featuresCount; ++i){
		ParameterCombo->addItem(features.GetOptionName(i), features.GetOptionId(i));
		ParameterCombo->setItemData(ParameterCombo->count()-1, features.GetOptionDescription(i), Qt::ToolTipRole);
	}
}


// protected methods

void CBlobFilterGui::UpdateComboBoxes()
{
	ConditionCombo->setItemData(0, iblob::IBlobFilterParams::FC_BETWEEN);
	ConditionCombo->setItemData(1, iblob::IBlobFilterParams::FC_OUTSIDE);
	ConditionCombo->setItemData(2, iblob::IBlobFilterParams::FC_EQUAL);
	ConditionCombo->setItemData(3, iblob::IBlobFilterParams::FC_NOT_EQUAL);
	ConditionCombo->setItemData(4, iblob::IBlobFilterParams::FC_GREATER);
	ConditionCombo->setItemData(5, iblob::IBlobFilterParams::FC_GREATER_EQUAL);
	ConditionCombo->setItemData(6, iblob::IBlobFilterParams::FC_LESS);
	ConditionCombo->setItemData(7, iblob::IBlobFilterParams::FC_LESS_EQUAL);
}


void CBlobFilterGui::UpdateFilter()
{
	iqt::CSignalBlocker block(ParameterCombo);

	// only show value2 for BETWEEN and OUTSIDE filters
	int condition = ConditionCombo->itemData(ConditionCombo->currentIndex()).toInt();
	bool showValue2 =
				(condition == iblob::IBlobFilterParams::FC_BETWEEN) ||
				(condition == iblob::IBlobFilterParams::FC_OUTSIDE);

	AndLabel->setVisible(showValue2);
	Value2SpinBox->setVisible(showValue2);
}


void CBlobFilterGui::OnFilterParameterChanged()
{
	UpdateFilter();

	emit FilterChanged();
}


void CBlobFilterGui::on_RemoveButton_clicked()
{
	emit RemoveFilter(this);
}


} // namespace iblobgui

